## Import library and load data

In [ ]:
from google.colab import drive
import pandas as pd
import numpy as np
import torch
import torch.nn as nn


In [ ]:
# Mount
drive.mount("/content/drive")

Mounted at /content/drive


In [ ]:
product_path = '/content/drive/MyDrive/BT4222 Group 3/1. Data Preparation/Data/product_metadata_parsed.parquet'
product_df = pd.read_parquet(product_path)
product_df.head()

wid                                               name  brand_id  \
0  90290292  89270446755918046025741917635585029871926...  75019956   
1  84294794  35314470278253305644862259056538602574191...  90358683   
2  43985934  99002012107475224750389807345717281391786...  99906610   
3  58357343  37269793566964962025051838249885733255404...  67566476   
4  71803457  53536632602574194190735260257419201104367...  78161242   

                            brand_name cate_id_1                 cate_name_1  \
0                    8927044675591804  75739187            176355875210311   
1                             56448622    920973  413810773160222581402210   
2                             99002012  24675518           2053319081620728   
3  37269793417877055669649659865766  98432384   49629523385795642329532   
4  53536632417877055360263159865766   6867541                    50134995   

  cate_id_2        cate_name_2 cate_id_3                          cate_name_3  \
0   9682321   176355863043105  56674307                             50298719   
1  72011611  4138107765056023  77105805                    2883009146686422   
2  45178979   976474541190979  41488568  10747522807345713385795688352935   
3  78943759           42329532  84880674                             25517595   
4  47765755   290298641345793  66118678                             41907352   

  cate_id_4                          cate_name_4   shop_id  \
0  65420542                             50298719   9644527   
1  14729718                    2883009146686422  51346695   
2   8059148  10747522807345713385795688352935  71737804   
3  16697032                             25517595  45062963   
4  70549743                             41907352  24505854   

                                           name_list  \
0  [89270446, 75591804, 60257419, 1763558, 502987...   
1  [35314470, 27825330, 56448622, 59056538, 60257...   
2  [99002012, 10747522, 4750389, 80734571, 728139...   
3  [37269793, 56696496, 20250518, 38249885, 73325...   
4  [53536632, 60257419, 41907352, 60257419, 20110...   

                            brand_name_list                cate_name_1_list  \
0                      [89270446, 75591804]             [1763558, 75210311]   
1                                [56448622]  [41381077, 31602225, 81402210]   
2                                [99002012]            [20533190, 81620728]   
3  [37269793, 41787705, 56696496, 59865766]   [4962952, 33857956, 42329532]   
4  [53536632, 41787705, 53602631, 59865766]                      [50134995]   

       cate_name_2_list                          cate_name_3_list  \
0   [1763558, 63043105]                                [50298719]   
1  [41381077, 65056023]                      [28830091, 46686422]   
2   [9764745, 41190979]  [10747522, 80734571, 33857956, 88352935]   
3            [42329532]                                [25517595]   
4   [2902986, 41345793]                                [41907352]   

                           cate_name_4_list  
0                                [50298719]  
1                      [28830091, 46686422]  
2  [10747522, 80734571, 33857956, 88352935]  
3                                [25517595]  
4                                [41907352]

In [ ]:
user_behaviour_path = '/content/drive/MyDrive/BT4222 Group 3/1. Data Preparation/Data/user_behavior_data_sampled_parsed_2.parquet'
user_behaviour_df = pd.read_parquet(user_behaviour_path)
user_behaviour_df.head()

query  \
0                63995226602574196100244625575117   
1                                  1254963427438884   
2                          24374376067768939206305   
3  29463948632095349356845164880503931105432...   
4                           2793914964365783663289   

                                  candidate_wid_list  \
0  [36098226, 82122191, 34768767, 71429327, 87635...   
1  [30857980, 44577431, 82696473, 94722703, 51139...   
2  [42155506, 90092899, 66647496, 92302381, 61552...   
3  [97234676, 83859812, 90043344, 51562568, 19973...   
4  [19711065, 83045333, 65555812, 23738187, 84489...   

                                candidate_label_list  \
0                          [1.0, 0.0, 0.0, 0.0, 0.0]   
1  [1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
2  [1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
3                          [1.0, 0.0, 0.0, 0.0, 0.0]   
4  [1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   

                                    history_qry_list  \
0  [-1, 68784030, 9592291398423633, 97447992, 97...   
1  [9755613572694219, 46916741, -1, 26082679551...   
2                                           [-1, -1]   
3  [3426234151020854598336931355834712920960,...   
4  [606776896025741971768678, 3989354160257419...   

                                    history_wid_list  \
0  [9227588, 87422901, 21056095, 65178325, 549785...   
1  [18256630, 46206270, 86878921, 82100823, 49359...   
2                                [8133925, 59478082]   
3  [49854616, 56947170, 16552251, 79406366, 73208...   
4  [45800470, 18419519, 30220497, 81499421, 52650...   

                                   history_type_list  \
0  [ORD, CART, CART, CART, CART, CART, CART, CART...   
1  [ORD, ORD, CART, CART, CART, CART, ORD, ORD, C...   
2                                     [CLICK, CLICK]   
3  [CLICK, CLICK, CLICK, CLICK, CLICK, CLICK, CLI...   
4  [ORD, ORD, ORD, ORD, ORD, ORD, ORD, ORD, ORD, ...   

                                   history_time_list  \
0  [0, 82448, 10315, 30, 33, 241, 18, 33, 3, 21, ...   
1  [0, 4813478, 14061983, 177223, 1304133, 42, 19...   
2                                            [0, 27]   
3  [0, 12, 4, 14, 6, 58095, 1524, 481, 20, 363, 7...   
4  [2015453, 958482, 13678, 1193864, 834215, 0, 0...   

         candidate_label_list_int  \
0                 [1, 0, 0, 0, 0]   
1  [1, 1, 0, 0, 0, 0, 0, 0, 0, 0]   
2  [1, 1, 0, 0, 0, 0, 0, 0, 0, 0]   
3                 [1, 0, 0, 0, 0]   
4  [1, 1, 0, 0, 0, 0, 0, 0, 0, 0]   

                           history_type_list_ordinal  \
0  [3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, ...   
1  [3.0, 3.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 2.0, ...   
2                                         [1.0, 1.0]   
3  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
4  [3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, ...   

                                          query_list  \
0           [63995226, 60257419, 61002446, 25575117]   
1                               [12549634, 27438884]   
2                      [2437437, 60677689, 39206305]   
3  [29463948, 63209534, 93568451, 64880503, 93110...   
4                       [2793914, 96436578, 3663289]   

                              history_qry_list_terms  
0  [[-1], [68784030], [95922913, 98423633], [9744...  
1  [[97556135, 72694219], [46916741], [-1], [2608...  
2                                       [[-1], [-1]]  
3  [[34262341, 51020854, 59833693, 13558347, 1292...  
4  [[60677689, 60257419, 71768678], [39893541, 60...

## Candidate Encoding

Encode categorical IDs in product_df
- assign integer indices to each unique ID (starting from 1, 0 = padding/UNK)
- the mappings are stored in a dictionary e.g. wid2idx = {84294794: 12, ...}

In [ ]:
# Create a copy (just in case)
prod = product_df.copy()

def encode_col(df, colname):
    """
    Map raw IDs in df[colname] to contiguous indices:
    0 reserved for PAD/UNK.
    Returns: index_series, rawID -> index dict
    """
    # Ensure we treat as integers where possible
    df[colname] = df[colname].astype("Int64")

    uniques = df[colname].dropna().unique()

    # Map each unique non-null raw ID to 1..N
    id2idx = {int(v): i+1 for i, v in enumerate(uniques)}

    # For each value in the column, try to find the new ID, if not 0
    def mapper(x):
        if pd.isna(x):
            return 0
        return id2idx.get(int(x), 0)  # unseen -> 0 (UNK)

    idx_series = df[colname].map(mapper).astype(int)
    return idx_series, id2idx

# 1) wid ("primary id")
prod["wid_idx"], wid2idx = encode_col(prod, "wid")

# 2) brand, shop, category levels
prod["brand_idx"], brand2idx = encode_col(prod, "brand_id")
prod["shop_idx"],  shop2idx  = encode_col(prod, "shop_id")

prod["cate1_idx"], cate1_2idx = encode_col(prod, "cate_id_1")
prod["cate2_idx"], cate2_2idx = encode_col(prod, "cate_id_2")
prod["cate3_idx"], cate3_2idx = encode_col(prod, "cate_id_3")
prod["cate4_idx"], cate4_2idx = encode_col(prod, "cate_id_4")


Align rows + build lookup tensors
- Given wid_idx, we want to quickly know its brand/shop/cate indices using tensors
- A lookup tensor is just a 1-D array where: the position = key (wid_idx) and the value = mapped attribute (brand_idx)
- brand_lookup[wid_idx] -> brand_idx, shop_lookup[wid_idx] -> shop_idx, etc.
- Row 0 is all zeros (PAD/UNK)
| wid_idx | brand_idx | shop_idx | cate1_idx | cate2_idx | cate3_idx | cate4_idx |               |
| ------- | --------- | -------- | --------- | --------- | --------- | --------- | ------------- |
| 0       | 0         | 0        | 0         | 0         | 0         | 0         |               |
| 1       | 11        | 3        | 7         | 2         | 9         | 1         |               |
| 2       | 4         | 8        | 2         | 7         | 5         | 4         |               |
| **12**  | **5**     | **9**    | **7**     | **2**     | **6**     | **3**     | ← our product |
| ...     | ...       | ...      | ...       | ...       | ...       | ...       |               |


In [ ]:
# Sort the dataframe by wid index (row 1 in the dataframe corresponds to wid_idx = 1, row 2 → wid_idx = 2, etc.)
prod = prod.sort_values("wid_idx").reset_index(drop=True)

# Vocabulary sizes / No. of unique IDs (+1 cause we reserved index 0 as a special “padding / unknown” ID)
num_wid   = int(prod["wid_idx"].max()) + 1
num_brand = int(prod["brand_idx"].max()) + 1
num_shop  = int(prod["shop_idx"].max()) + 1
num_c1    = int(prod["cate1_idx"].max()) + 1
num_c2    = int(prod["cate2_idx"].max()) + 1
num_c3    = int(prod["cate3_idx"].max()) + 1
num_c4    = int(prod["cate4_idx"].max()) + 1

# Initialise lookup tables; index = wid_idx; value = corresponding feature idx
brand_lookup = torch.zeros(num_wid, dtype=torch.long)
shop_lookup  = torch.zeros(num_wid, dtype=torch.long)
c1_lookup    = torch.zeros(num_wid, dtype=torch.long)
c2_lookup    = torch.zeros(num_wid, dtype=torch.long)
c3_lookup    = torch.zeros(num_wid, dtype=torch.long)
c4_lookup    = torch.zeros(num_wid, dtype=torch.long)

for _, row in prod.iterrows():
    w = int(row["wid_idx"])
    if w == 0:
        continue  # padding / unk
    brand_lookup[w] = int(row["brand_idx"])
    shop_lookup[w]  = int(row["shop_idx"])
    c1_lookup[w]    = int(row["cate1_idx"])
    c2_lookup[w]    = int(row["cate2_idx"])
    c3_lookup[w]    = int(row["cate3_idx"])
    c4_lookup[w]    = int(row["cate4_idx"])


Map candidate_wid_list → wid_idx in user_behaviour_df
- [36098226, 82122191, 34768767, 71429327, 87635] → [15, 7, 12, 23, 19]

In [ ]:
# Create a copy (just in case)
ub = user_behaviour_df.copy()

def map_wids_to_idx(wid_list):
    safe_indices = []
    for w in wid_list:
        # Skip if missing or invalid
        if pd.isna(w) or w == 'nan' or w is None:
            safe_indices.append(0)  # 0 = padding/unknown
            continue
        try:
            wid_int = int(w)
        except (ValueError, TypeError):
            wid_int = 0
        safe_indices.append(wid2idx.get(wid_int, 0))
    return safe_indices

ub["candidate_wid_idx_list"] = ub["candidate_wid_list"].apply(map_wids_to_idx)

Candidate Encoder module
- For every candidate product (wid_idx), it:
  -  Looks up which brand, shop, and category levels it belongs to
  - Converts each of those categorical IDs into a dense vector (embedding)
  - Combines (sums) them into a single 64-dimensional candidate embedding vector
- Basically, given a product ID, give me a vector that summarizes all its attributes

In [ ]:
class CandidateEncoder(nn.Module):
    def __init__(self,
                 num_wid,
                 num_brand,
                 num_shop,
                 num_c1,
                 num_c2,
                 num_c3,
                 num_c4,
                 brand_lookup,
                 shop_lookup,
                 c1_lookup,
                 c2_lookup,
                 c3_lookup,
                 c4_lookup,
                 emb_dim=64):
        super().__init__()
        self.emb_dim = emb_dim

        # Embedding tables (0 index = padding/UNK)
        self.wid_emb   = nn.Embedding(num_wid,   emb_dim, padding_idx=0) # matrix of shape (num_wid, emb_dim = 64)
        self.brand_emb = nn.Embedding(num_brand, emb_dim, padding_idx=0)
        self.shop_emb  = nn.Embedding(num_shop,  emb_dim, padding_idx=0)
        self.c1_emb    = nn.Embedding(num_c1,    emb_dim, padding_idx=0)
        self.c2_emb    = nn.Embedding(num_c2,    emb_dim, padding_idx=0)
        self.c3_emb    = nn.Embedding(num_c3,    emb_dim, padding_idx=0)
        self.c4_emb    = nn.Embedding(num_c4,    emb_dim, padding_idx=0)

        # Register lookups as buffers (move with model.to(device), not trainable)
        self.register_buffer("brand_lookup", brand_lookup)
        self.register_buffer("shop_lookup",  shop_lookup)
        self.register_buffer("c1_lookup",    c1_lookup)
        self.register_buffer("c2_lookup",    c2_lookup)
        self.register_buffer("c3_lookup",    c3_lookup)
        self.register_buffer("c4_lookup",    c4_lookup)

    def forward(self, candidate_wid_idx):
        """
        candidate_wid_idx: LongTensor (B, C)
        Each entry is wid_idx (0 allowed as PAD).
        Returns: candidate_emb: (B, C, emb_dim)
        """
        wid_idx = candidate_wid_idx  # (B = number of query sessions, C = number of candidates per query)

        # Look up attribute indices for each wid_idx
        brand_idx = self.brand_lookup[wid_idx]  # (B, C)
        shop_idx  = self.shop_lookup[wid_idx]
        c1_idx    = self.c1_lookup[wid_idx]
        c2_idx    = self.c2_lookup[wid_idx]
        c3_idx    = self.c3_lookup[wid_idx]
        c4_idx    = self.c4_lookup[wid_idx]

        # Embed each attribute: all -> (B, C, emb_dim)
        wid_e   = self.wid_emb(wid_idx)
        brand_e = self.brand_emb(brand_idx)
        shop_e  = self.shop_emb(shop_idx)
        c1_e    = self.c1_emb(c1_idx)
        c2_e    = self.c2_emb(c2_idx)
        c3_e    = self.c3_emb(c3_idx)
        c4_e    = self.c4_emb(c4_idx)

        # Sum to get final candidate embedding (element wise sum)
        cand_emb = wid_e + brand_e + shop_e + c1_e + c2_e + c3_e + c4_e

        return cand_emb


In [ ]:
# Instantiate it
emb_dim = 64  # TBC NEEDS TO BE AGREED

cand_encoder = CandidateEncoder(
    num_wid=num_wid,
    num_brand=num_brand,
    num_shop=num_shop,
    num_c1=num_c1,
    num_c2=num_c2,
    num_c3=num_c3,
    num_c4=num_c4,
    brand_lookup=brand_lookup,
    shop_lookup=shop_lookup,
    c1_lookup=c1_lookup,
    c2_lookup=c2_lookup,
    c3_lookup=c3_lookup,
    c4_lookup=c4_lookup,
    emb_dim=emb_dim,
)

Example use
- the forward expects a (B, C) tensor but for every user, they have candidate wid idx list of vraiable lengths. since max is 200, we pad everyone with 0 until C = 200

In [ ]:
from torch.nn.utils.rnn import pad_sequence

# Convert each candidate list to a tensor
candidate_tensors = [
    torch.tensor(lst, dtype=torch.long)
    for lst in ub["candidate_wid_idx_list"]
]

# Pad them to equal length (max length = 200)
candidate_padded = pad_sequence(candidate_tensors, batch_first=True, padding_value=0)

# candidate_padded shape = (173831, 200)
print(candidate_padded.shape)

# Move both model and data to device (GPU if available)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
cand_encoder = cand_encoder.to(device)
candidate_padded = candidate_padded.to(device)


# Forward pass: get candidate embeddings
with torch.no_grad():  # disable gradient calc for inference/embedding extraction
    candidate_embeddings = cand_encoder(candidate_padded)

# Get embedding for a wid
wid = 84294794
wid_idx = wid2idx.get(wid, 0)

with torch.no_grad():
    vector = cand_encoder.wid_emb.weight[wid_idx]  # (64,)

vector.shape


torch.Size([173831, 200])


## Save full encoder + mappings

In [ ]:
save_path = '/content/drive/MyDrive/BT4222 Group 3/2. Models/Encoders/candidate_encoder_ckpt.pt'

checkpoint = {
    "state_dict": cand_encoder.state_dict(),
    "emb_dim": emb_dim,
    "wid2idx": wid2idx,
    "brand2idx": brand2idx,
    "shop2idx": shop2idx,
    "cate1_2idx": cate1_2idx,
    "cate2_2idx": cate2_2idx,
    "cate3_2idx": cate3_2idx,
    "cate4_2idx": cate4_2idx,
    "brand_lookup": brand_lookup,
    "shop_lookup": shop_lookup,
    "c1_lookup": c1_lookup,
    "c2_lookup": c2_lookup,
    "c3_lookup": c3_lookup,
    "c4_lookup": c4_lookup,
    "num_wid": num_wid,
    "num_brand": num_brand,
    "num_shop": num_shop,
    "num_c1": num_c1,
    "num_c2": num_c2,
    "num_c3": num_c3,
    "num_c4": num_c4,
}

torch.save(checkpoint, save_path)
print(f"Saved candidate encoder to {save_path}")